identity-iota/src/tangle/network.rs

```diff
@@ -16,22 +19,29 @@ lazy_static! {
 }
 
 /// The Tangle network to use (`Mainnet` or `Testnet`).
-#[derive(Clone, Copy, Debug, Hash, PartialEq, Eq, PartialOrd, Ord, Deserialize, Serialize)]
+#[derive(Clone, Debug, Hash, PartialEq, Eq, PartialOrd, Ord, Deserialize, Serialize)]
 pub enum Network {
   #[serde(rename = "main")]
   Mainnet,
   #[serde(rename = "test")]
   Testnet,
+  Other(String),
 }
 
 impl Network {
-  /// Parses the provided string to a `Network`.
+  /// Parses the provided string to a [`Network`].
+  ///
+  /// The inputs `"test"` and `"main"` will be mapped to the well-known [`Testnet`][Network::Testnet]
+  /// and [`Mainnet`][Network::Mainnet] variants respectively.
+  /// Other inputs will return an instance of [`Other`][Network::Other].
   ///
-  /// If the input is `"test"` then `Testnet` is returned, otherwise `Mainnet` is returned.
-  pub fn from_name(string: &str) -> Self {
+  /// Note that empty strings are not valid network names.
+  pub fn from_name(string: &str) -> Result<Self> {
     match string {
-      TEST_NETWORK_NAME => Self::Testnet,
-      _ => Self::Mainnet,
+      "" => Err(Error::InvalidDIDNetwork),
+      TEST_NETWORK_NAME => Ok(Self::Testnet),
+      MAIN_NETWORK_NAME => Ok(Self::Mainnet),
+      other => Ok(Self::Other(other.to_owned())),
     }
   }
```

Comment:
We should probably enforce that the network name adheres to the IOTA DID Spec:
https://github.com/iotaledger/identity.rs/blob/99049c0e4a64a7babf224298d5d38bc243421965/documentation/docs/specs/iota_did_method_spec.md

That is, completely lowercase, only chars in `{0-9,a-b}`, and only up to 6 chars (is that arbitrary or a hard limit?). We can maybe transform it in `from_name` to lowercase, strip out spaces etc. but some basic validation to prevent DIDs being generated with invalid network tags seems necessary at the very least.

Reply:
You raise a good point!

The name of the `Network`, however, isn't necessarily the same as the iota-network-id . It's possible to publish a `did:iota:arbitrary_string:...` to the [testnet](https://explorer.iota.org/testnet/message/f290987610dbbb251193b103cbf125befb5006eda89688bdc89a1fd7dff94683). This is because this method takes an arbitrary string, not a `Network` type: `IotaDocument::from_keypair_with_network(keypair, "arbitrary_string");` So the spec isn't currently enforced in the low-level API. In current `dev`, using the Account API with `IdentityCreate::new().network("arbitrary")` also lets users publish a `did:iota:arbitrary:...` to the [mainnet](https://explorer.iota.org/mainnet/message/941f463dea07dcb7b145ee0229dca5f6ac091a8fca0398413ea4d23a946a93ab). 

Which raises the question, why isn't this enforced currently, and do we want to change that? I guess so. In that case, we shouldn't limit the name of the `Network` to 6 characters, or even just `{0-9,a-b}`. If someone has a `private-tangle` running, they probably don't want to change the name of the network just to publish their did. Then the reasonable thing to do, is to create a separate type that enforces the spec.

Reply:
What @cycraig said is correct. It should enforce the method spec with 6 chars, lowercase and `{0-9,a-z}`. The "name of the network" for the identity framework indeed should mean the iota-network-id, any other names are arbitrary and meaningless to the framework. Btw, the 6-char limit is arbitrary, but something had to be chosen.